File: q126-150/q146.py
# 2020.03.23
# Problem Statement:
# https://leetcode.com/problems/lru-cache/
# Could not think of an O(n) solution...
# https://leetcode.com/problems/lru-cache/discuss/45926/Python-Dict-%2B-Double-LinkedList

# double linked list
class Node:
    def __init__(self, key=0, val=0, prev=None, next=None):
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next
        
class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.dict = {} # value would be the entire corresponding node
        self.keys_head = Node()
        self.keys_tail = Node(prev=self.keys_head)
        self.keys_head.next = self.keys_tail

    # remove from current location  
    def remove(self, node):
        prev = node.prev
        next = node.next
        prev.next = next
        next.prev = prev
    
    # add to the tail
    def add(self, node):
        prev = self.keys_tail.prev
        prev.next = node
        node.prev = prev
        node.next = self.keys_tail
        self.keys_tail.prev = node
        
    def get(self, key: int) -> int:
        if key not in self.dict.keys(): return -1
        else:
            # remove from current location
            # add it before keys_tail
            node_to_change = self.dict[key]
            self.remove(node_to_change)            
            self.add(node_to_change)
            return self.dict[key].val
        

    def put(self, key: int, value: int) -> None:
        if key not in self.dict.keys():
            new_node = Node(key, value)
            self.add(new_node)
            self.dict[key] = new_node
            
            # remove the first element in the linked list
            if len(self.dict) > self.capacity:
                node_to_remove = self.keys_head.next
                self.remove(node_to_remove)
                del self.dict[node_to_remove.key]               
                
        else:
            # change the value
            self.dict[key].val = value
            # remove from current location
            # add it before keys_tail
            node_to_change = self.dict[key]
            self.remove(node_to_change)
            self.add(node_to_change)
            
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: Amazon/q146.java
// 2022.05.17
// Problem Statement:
// https://leetcode.com/problems/lru-cache/
// did it in 2021/03/23 in python

// idea: double linked list, otherwise TLE
// also have to keep head and tail together for faster add_to_tail operations
class LRUCache {
    public int cap;
    class DNode {
        int key;
        int value;
        DNode left;
        DNode right;
    }
    public DNode head;
    public DNode tail;
    public Map <Integer, DNode> ht;

    public void removeNode (DNode node) {
        DNode l_node = node.left;
        DNode r_node = node.right;
        if (l_node==null && r_node==null) {
            head = null;
            tail = null;
            return;
        } else if (l_node==null) {
            head = head.right;
            node.right = null;
            r_node.left = null;
            return;
        } else if (r_node==null) {
            tail = tail.left;
            l_node.right = null;
            node.left = null;
            return;
        } else {
            node.left = null;
            node.right = null;
            l_node.right = r_node;
            r_node.left = l_node;
            return;
        }
    }
    
    public void addNode (DNode node) {
        if (head==null) { // empty list
            head = node;
            tail = node;
            return;
        } else {
            tail.right = node;
            node.left = tail;
            tail = tail.right;
            return;
        }
    } 
    
    public LRUCache(int capacity) {
        cap = capacity;
        //list = new LinkedList<Integer> ();
        head = null;
        tail = null;
        ht = new HashMap <Integer, DNode> ();
    }
    
    public int get(int key) {
        if (ht.containsKey(key)) {
            removeNode(ht.get(key)); // remove from the middle
            addNode(ht.get(key)); // attach to the tail
            return ht.get(key).value;
        } else {
            return -1;
        }
    }
    
    public void put(int key, int value) {
        if (ht.containsKey(key)) {
            ht.get(key).value = value;
            ht.put(key, ht.get(key));
            removeNode(ht.get(key)); // remove from the middle
            addNode(ht.get(key)); // attach to the tail
        } else {
            if (ht.size()<cap) { // direct add
                DNode node = new DNode ();
                node.key = key;
                node.value = value;
                ht.put(key, node);
                addNode(node);
            } else { // have to remove the least recent used node
                DNode node = new DNode ();
                node.key = key;
                node.value = value;
                ht.put(key, node);
                ht.remove(head.key); // remove from ht
                removeNode(head); // remove from linked list
                addNode(node);
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */